File: client/go/group.go
package picomq

import (
	"context"
	"errors"
	"math"
	"sync"
	"time"
)

type GroupConfig struct {
	SessionTimeout    time.Duration
	HeartbeatInterval time.Duration
	RebalanceTimeout  time.Duration
	InstanceID        string
	ClientID          string
	Retry             RetryPolicy
}

func DefaultGroupConfig() GroupConfig {
	return GroupConfig{SessionTimeout: 30 * time.Second, Retry: RetryPolicy{MaxAttempts: math.MaxInt32, InitialBackoff: 100 * time.Millisecond, MaxBackoff: 5 * time.Second, Multiplier: 2}}
}

func (c GroupConfig) joinOptions(memberID string) JoinOptions {
	return JoinOptions{MemberID: memberID, InstanceID: c.InstanceID, ClientID: c.ClientID, SessionTimeout: c.SessionTimeout, RebalanceTimeout: c.RebalanceTimeout}
}

type Assignment struct {
	Generation int32
	Streams    []string
}

type GroupMember struct {
	client       *PicoClient
	group        string
	subscription []string
	config       GroupConfig
	assignments  chan Assignment
	cancel       context.CancelFunc
	done         chan struct{}
	mu           sync.Mutex
	memberID     string
	generation   int32
	current      Assignment
	err          error
}

func (c *PicoClient) NewGroupMember(ctx context.Context, group string, subscription []string, config *GroupConfig) (*GroupMember, error) {
	cfg := DefaultGroupConfig()
	if config != nil {
		cfg = *config
		if cfg.SessionTimeout <= 0 {
			cfg.SessionTimeout = 30 * time.Second
		}
		if cfg.Retry.MaxAttempts == 0 {
			cfg.Retry = DefaultGroupConfig().Retry
		}
	}
	if cfg.HeartbeatInterval <= 0 {
		cfg.HeartbeatInterval = cfg.SessionTimeout / 3
	}
	if cfg.HeartbeatInterval < time.Millisecond {
		cfg.HeartbeatInterval = time.Millisecond
	}
	m := &GroupMember{client: c, group: group, subscription: append([]string(nil), subscription...), config: cfg, assignments: make(chan Assignment, 1), done: make(chan struct{})}
	joined, err := m.rejoin(ctx, "")
	if err != nil {
		return nil, err
	}
	m.apply(joined)
	loopCtx, cancel := context.WithCancel(context.Background())
	m.cancel = cancel
	go m.run(loopCtx)
	return m, nil
}

func (m *GroupMember) MemberID() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.memberID
}

func (m *GroupMember) Assignment() Assignment {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.current
}

func (m *GroupMember) Assignments() <-chan Assignment { return m.assignments }

func (m *GroupMember) Err() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.err
}

func (m *GroupMember) Commit(ctx context.Context, offsets Offsets) error {
	m.mu.Lock()
	if m.err != nil {
		defer m.mu.Unlock()
		return m.err
	}
	fence := m.fence()
	m.mu.Unlock()
	return m.client.CommitOffsets(ctx, m.group, offsets, &fence)
}

func (m *GroupMember) FetchOffsets(ctx context.Context, streams []string) (Offsets, error) {
	return m.client.FetchOffsets(ctx, m.group, streams)
}

func (m *GroupMember) Leave(ctx context.Context) error {
	m.cancel()
	select {
	case <-m.done:
	case <-ctx.Done():
		return ctx.Err()
	}
	m.mu.Lock()
	failed, memberID := m.err, m.memberID
	m.mu.Unlock()
	if failed != nil {
		return nil
	}
	return m.client.LeaveGroup(ctx, m.group, memberID, m.config.InstanceID)
}

func (m *GroupMember) fence() MemberFence {
	return MemberFence{MemberID: m.memberID, Generation: m.generation, InstanceID: m.config.InstanceID}
}

func (m *GroupMember) run(ctx context.Context) {
	defer close(m.done)
	defer close(m.assignments)
	attempt := 0
	for {
		if sleepCtx(ctx, m.config.HeartbeatInterval) != nil {
			return
		}
		m.mu.Lock()
		fence := m.fence()
		m.mu.Unlock()
		err := m.client.Heartbeat(ctx, m.group, fence)
		if err == nil {
			attempt = 0
			continue
		}
		if ctx.Err() != nil {
			return
		}
		var joined GroupMembership
		switch code := errorCode(err); {
		case code == "rebalance_in_progress" || code == "illegal_generation":
			joined, err = m.rejoin(ctx, fence.MemberID)
		case code == "unknown_member":
			joined, err = m.rejoin(ctx, "")
		case retryable(err):
			delay, again := m.config.Retry.delay(attempt)
			if again {
				attempt++
				if sleepCtx(ctx, delay) != nil {
					return
				}
				continue
			}
		}
		if ctx.Err() != nil {
			return
		}
		if err != nil {
			m.mu.Lock()
			m.err = err
			m.mu.Unlock()
			return
		}
		attempt = 0
		m.apply(joined)
	}
}

func (m *GroupMember) apply(joined GroupMembership) {
	m.mu.Lock()
	m.memberID, m.generation = joined.MemberID, joined.Generation
	m.current = Assignment{Generation: joined.Generation, Streams: joined.Assignment}
	current := m.current
	m.mu.Unlock()
	select {
	case <-m.assignments:
	default:
	}
	m.assignments <- current
}

func (m *GroupMember) rejoin(ctx context.Context, memberID string) (GroupMembership, error) {
	attempt := 0
	for {
		joined, err := m.client.JoinGroup(ctx, m.group, m.subscription, m.config.joinOptions(memberID))
		if err == nil {
			return joined, nil
		}
		if errorCode(err) == "unknown_member" {
			memberID = ""
			continue
		}
		if !retryable(err) {
			return GroupMembership{}, err
		}
		delay, again := m.config.Retry.delay(attempt)
		if !again {
			return GroupMembership{}, err
		}
		attempt++
		if err := sleepCtx(ctx, delay); err != nil {
			return GroupMembership{}, err
		}
	}
}

func errorCode(err error) string {
	var target *ClientError
	if errors.As(err, &target) {
		return target.Code
	}
	return ""
}

File: client/typescript/src/group.ts
import { asClientError, ClientError, throwIfAborted } from './error'
import { retryableError, sleep } from './util'
import { RetryPolicy } from './retry'
import type {
  CallOptions,
  GroupMembership,
  JoinOptions,
  MemberFence,
  Offsets,
} from './types'

const DEFAULT_SESSION_TIMEOUT_MS = 30_000

export interface GroupConfig {
  sessionTimeoutMs?: number
  heartbeatIntervalMs?: number
  rebalanceTimeoutMs?: number
  instanceId?: string
  clientId?: string
  retry?: RetryPolicy
  signal?: AbortSignal
}

export interface GroupClient {
  joinGroup(group: string, subscription: string[], options?: JoinOptions): Promise<GroupMembership>
  heartbeat(group: string, fence: MemberFence, options?: CallOptions): Promise<void>
  leaveGroup(
    group: string,
    memberId: string,
    instanceId?: string,
    options?: CallOptions,
  ): Promise<void>
  commitOffsets(
    group: string,
    offsets: Offsets,
    fence?: MemberFence,
    options?: CallOptions,
  ): Promise<void>
  fetchOffsets(group: string, streams?: string[], options?: CallOptions): Promise<Offsets>
}

export interface Assignment {
  generation: number
  streams: string[]
}

interface ResolvedConfig {
  sessionTimeoutMs: number
  heartbeatIntervalMs: number
  rebalanceTimeoutMs?: number
  instanceId?: string
  clientId?: string
  retry: RetryPolicy
}

function resolveConfig(config: GroupConfig): ResolvedConfig {
  const sessionTimeoutMs = config.sessionTimeoutMs ?? DEFAULT_SESSION_TIMEOUT_MS
  const resolved: ResolvedConfig = {
    sessionTimeoutMs,
    heartbeatIntervalMs: Math.max(1, config.heartbeatIntervalMs ?? sessionTimeoutMs / 3),
    retry: config.retry ?? new RetryPolicy(Number.MAX_SAFE_INTEGER, 100, 5_000, 2),
  }

  if (config.rebalanceTimeoutMs !== undefined) resolved.rebalanceTimeoutMs = config.rebalanceTimeoutMs
  if (config.instanceId !== undefined) resolved.instanceId = config.instanceId
  if (config.clientId !== undefined) resolved.clientId = config.clientId

  return resolved
}

function joinOptions(
  config: ResolvedConfig,
  memberId: string | undefined,
  signal: AbortSignal | undefined,
): JoinOptions {
  const options: JoinOptions = { sessionTimeoutMs: config.sessionTimeoutMs }

  if (signal !== undefined) options.signal = signal
  if (memberId !== undefined) options.memberId = memberId
  if (config.instanceId !== undefined) options.instanceId = config.instanceId
  if (config.clientId !== undefined) options.clientId = config.clientId
  if (config.rebalanceTimeoutMs !== undefined) options.rebalanceTimeoutMs = config.rebalanceTimeoutMs

  return options
}

class Watch<T> {
  private version = 0
  private waiters: Array<() => void> = []
  private closed = false

  constructor(private current: T) {}

  get value(): T {
    return this.current
  }

  set(value: T): void {
    this.current = value
    this.version += 1
    this.wake()
  }

  close(): void {
    this.closed = true
    this.wake()
  }

  async *changes(): AsyncGenerator<T> {
    let seen = -1
    for (;;) {
      if (this.version !== seen) {
        seen = this.version
        yield this.current
      }

      if (this.closed) return

      await new Promise<void>((resolve) => this.waiters.push(resolve))
    }
  }

  private wake(): void {
    const waiters = this.waiters
    this.waiters = []
    for (const waiter of waiters) waiter()
  }
}

export class GroupMember {
  private readonly controller = new AbortController()
  private readonly watch: Watch<Assignment>
  private readonly heartbeats: Promise<void>
  private currentMemberId: string
  private generation: number
  private failed: ClientError | undefined

  private constructor(
    private readonly client: GroupClient,
    private readonly group: string,
    private readonly subscription: string[],
    private readonly config: ResolvedConfig,
    joined: GroupMembership,
    signal: AbortSignal | undefined,
  ) {
    this.currentMemberId = joined.memberId
    this.generation = joined.generation
    this.watch = new Watch({ generation: joined.generation, streams: joined.assignment })

    if (signal !== undefined) {
      signal.addEventListener('abort', () => this.controller.abort(signal.reason), { once: true })
    }

    this.heartbeats = this.run()
  }

  static async join(
    client: GroupClient,
    group: string,
    subscription: string[],
    config: GroupConfig = {},
  ): Promise<GroupMember> {
    throwIfAborted(config.signal)
    const resolved = resolveConfig(config)

    const joined = await resolved.retry.run(
      () => client.joinGroup(group, subscription, joinOptions(resolved, undefined, config.signal)),
      retryableError,
      config.signal,
    )

    return new GroupMember(client, group, subscription, resolved, joined, config.signal)
  }

  memberId(): string {
    return this.currentMemberId
  }

  assignment(): Assignment {
    return this.watch.value
  }

  assignments(): AsyncIterable<Assignment> {
    return this.watch.changes()
  }

  error(): ClientError | undefined {
    return this.failed
  }

  async commit(offsets: Offsets, options?: CallOptions): Promise<void> {
    if (this.failed !== undefined) throw this.failed
    await this.client.commitOffsets(this.group, offsets, this.fence(), options)
  }

  async fetchOffsets(streams: string[] = [], options?: CallOptions): Promise<Offsets> {
    return this.client.fetchOffsets(this.group, streams, options)
  }

  async leave(options?: CallOptions): Promise<void> {
    this.controller.abort()
    await this.heartbeats
    this.watch.close()

    if (this.failed !== undefined) return

    await this.client.leaveGroup(this.group, this.currentMemberId, this.config.instanceId, options)
  }

  private fence(): MemberFence {
    const fence: MemberFence = { memberId: this.currentMemberId, generation: this.generation }
    if (this.config.instanceId !== undefined) fence.instanceId = this.config.instanceId
    return fence
  }

  private async run(): Promise<void> {
    const signal = this.controller.signal
    let attempt = 0

    for (;;) {
      try {
        await sleep(this.config.heartbeatIntervalMs, signal)
      } catch {
        return
      }

      const fence = this.fence()
      let joined: GroupMembership

      try {
        await this.client.heartbeat(this.group, fence, { signal })
        attempt = 0
        continue
      } catch (error) {
        const err = asClientError(error)
        if (err.kind === 'aborted') return

        try {
          if (err.code === 'rebalance_in_progress' || err.code === 'illegal_generation') {
            joined = await this.rejoin(fence.memberId, signal)
          } else if (err.code === 'unknown_member') {
            joined = await this.rejoin(undefined, signal)
          } else if (err.retryable()) {
            const delay = this.config.retry.delay(attempt)
            if (delay === null) throw err

            attempt += 1
            await sleep(delay, signal)
            continue
          } else {
            throw err
          }
        } catch (failure) {
          const fatal = asClientError(failure)
          if (fatal.kind === 'aborted') return

          this.failed = fatal
          this.watch.close()
          return
        }
      }

      attempt = 0
      this.currentMemberId = joined.memberId
      this.generation = joined.generation
      this.watch.set({ generation: joined.generation, streams: joined.assignment })
    }
  }

  private async rejoin(
    memberId: string | undefined,
    signal: AbortSignal,
  ): Promise<GroupMembership> {
    let attempt = 0

    for (;;) {
      try {
        return await this.client.joinGroup(
          this.group,
          this.subscription,
          joinOptions(this.config, memberId, signal),
        )
      } catch (error) {
        const err = asClientError(error)

        if (err.code === 'unknown_member') {
          memberId = undefined
          continue
        }

        if (err.retryable()) {
          const delay = this.config.retry.delay(attempt)
          if (delay === null) throw err

          attempt += 1
          await sleep(delay, signal)
          continue
        }

        throw err
      }
    }
  }
}
